spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestConflictValidation.java

```diff
@@ -0,0 +1,121 @@
+    sql("INSERT INTO %s VALUES (1, 'a'), (2, 'b'), (3, 'c')", tableName);
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+  }
+
+  @Test
+  public void testDynamicValidation() throws Exception {
+    Table table = validationCatalog.loadTable(tableIdent);
+    long snapshotId = table.currentSnapshot().snapshotId();
+
+    List<SimpleRecord> records = Lists.newArrayList(
+        new SimpleRecord(1, "a"));
+    Dataset<Row> df = spark.createDataFrame(records, SimpleRecord.class);
+    df.select("id", "data").writeTo(tableName).append();
+
+    // Validating from previous snapshot finds conflicts
+    Dataset<Row> conflictingDf = spark.createDataFrame(records, SimpleRecord.class);
+    Exception thrown = null;
+    try {
+      conflictingDf.writeTo(tableName)
+          .option(SparkWriteOptions.VALIDATE_FROM_SNAPSHOT_ID, String.valueOf(snapshotId))
+          .option(SparkWriteOptions.DYNAMIC_OVERWRITE_ISOLATION_LEVEL, IsolationLevel.SERIALIZABLE.toString())
+          .overwritePartitions();
+    } catch (Exception e) {
+      thrown = e;
+    }
+    Assert.assertNotNull("Expected validation exception but none thrown", thrown);
+    Assert.assertTrue("Nested validation exception", thrown.getCause() instanceof ValidationException);
+    Assert.assertTrue(thrown.getCause().getMessage().contains(
+        "Found conflicting files that can contain records matching partitions [id=1]"));
+
+    // Validating from latest snapshot should succeed
+    table.refresh();
+    snapshotId = table.currentSnapshot().snapshotId();
+    conflictingDf.writeTo(tableName)
+        .option(SparkWriteOptions.VALIDATE_FROM_SNAPSHOT_ID, String.valueOf(snapshotId))
+        .option(SparkWriteOptions.DYNAMIC_OVERWRITE_ISOLATION_LEVEL, IsolationLevel.SERIALIZABLE.toString())
+        .overwritePartitions();
+  }
+
+  @Test
+  public void testDynamicValidationNoSnapshotId() throws Exception {
+    Table table = validationCatalog.loadTable(tableIdent);
+
+    List<SimpleRecord> records = Lists.newArrayList(
+        new SimpleRecord(1, "a"));
+    Dataset<Row> df = spark.createDataFrame(records, SimpleRecord.class);
+    df.select("id", "data").writeTo(tableName).append();
+
+    // Validating from previous snapshot finds conflicts
+    Dataset<Row> conflictingDf = spark.createDataFrame(records, SimpleRecord.class);
+    Exception thrown = null;
+    try {
+      conflictingDf.writeTo(tableName)
+          .option(SparkWriteOptions.DYNAMIC_OVERWRITE_ISOLATION_LEVEL, IsolationLevel.SERIALIZABLE.toString())
+          .overwritePartitions();
+    } catch (Exception e) {
+      thrown = e;
+    }
+    Assert.assertNotNull("Expected validation exception but none thrown", thrown);
+    Assert.assertTrue("Nested validation exception", thrown.getCause() instanceof ValidationException);
+    Assert.assertTrue(thrown.getCause().getMessage().contains(
+        "Cannot determine history between starting snapshot"));
```

Comment:
@aokolnychyi after changing this path to use the common method MergingSnapshotProducer::validateAddedData/DeleteFiles, it will throw this kind of exception if user sets serializable-isolation but not the validate-from-snapshot (common method tries to get the history and it fails as there is no startingSnapshotId). Is this a good expected behavior?

Reply:
I think it is related to 0 snapshot ID used as default for `validate-from-snapshot-id`. I hope if we fix the default value, it will fix the error message.

Reply:
You are right, this worked.